src/librbd/Group.cc

```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+			  snap_seq;
+  return ind_snap_name_stream.str();
+}
+
+int group_snap_create(librados::IoCtx& group_ioctx,
+		      const char *group_name, const char *snap_name)
+{
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  librados::Rados rados(group_ioctx);
+
+  int r = group_snap_name_check_duplicate(group_ioctx, group_name, snap_name);
+  if (r < 0) {
+    return r;
+  }
+
+  string group_id;
+  uint64_t snap_seq = 0;
+  cls::rbd::GroupSnapshot gs;
+  vector<cls::rbd::ImageSnapshotRef> image_snaps;
+  std::string ind_snap_name;
+
+  r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
+			     group_name, &group_id);
+  if (r < 0) {
+    lderr(cct) << "error reading consistency group id object: "
+	       << cpp_strerror(r)
+	       << dendl;
+    return r;
+  }
+  string group_header_oid = util::group_header_name(group_id);
+
+  std::vector<group_image_status_t> images;
+  r = librbd::group_image_list(group_ioctx, group_name, images);
+  if (r < 0) {
+    return r;
+  }
+  int n = images.size();
+  std::vector<librbd::IoCtx*> io_ctxs;
+  std::vector<librbd::ImageCtx*> ictxs;
+  std::vector<C_SaferCond*> on_finishes;
+  for (auto i: images) {
+    librbd::IoCtx* image_io_ctx = new librbd::IoCtx;
+
+    r = rados.ioctx_create2(i.pool, *image_io_ctx);
+    if (r < 0) {
+      ldout(cct, 1) << "Failed to create io context for image" << dendl;
+    }
+
+    librbd::ImageCtx* image_ctx = new ImageCtx(i.name.c_str(), "", "",
+					       *image_io_ctx, false);
+
+    C_SaferCond* on_finish = new C_SaferCond;
+
+    image_ctx->state->open(on_finish);
+
+    ictxs.push_back(image_ctx);
+    on_finishes.push_back(on_finish);
+  }
+  int ret_code = 0;
+  for (int i = 0; i < n; ++i) {
+    r = on_finishes[i]->wait();
+    delete on_finishes[i];
+    if (r < 0) {
+      delete ictxs[i];
+      ictxs[i] = nullptr;
+      ret_code = r;
+    }
+  }
+  if (ret_code != 0) {
+    goto cleanup;
+  }
+  for (auto i: ictxs) {
+    i->exclusive_lock->block_requests(-EBUSY);
+  }
+  for (int i = 0; i < n; ++i) {
+    ImageCtx *ictx = ictxs[i];
+    RWLock::RLocker owner_lock(ictx->owner_lock);
+
+    on_finishes[i] = new C_SaferCond;
+    ictx->exclusive_lock->request_lock(on_finishes[i]);
+  }
+
+  ret_code = 0;
+  for (auto i: on_finishes) {
+    r = i->wait();
+    delete i;
+    if (r < 0) {
+      ret_code = r;
+    }
+  }
+  if (ret_code != 0) {
+    goto cleanup;
+  }
+
+  r = cls_client::group_snap_next_seq(&group_ioctx, group_header_oid, &snap_seq);
+  if (r < 0) {
+    ret_code = r;
+    goto cleanup;
+  }
+
+  gs.id = snap_seq;
+  gs.uuid = "";
+  gs.name = string(snap_name);
+  gs.state = cls::rbd::GROUP_SNAPSHOT_STATE_PENDING;
+
+  r = cls_client::group_snap_save(&group_ioctx, group_header_oid, gs);
+  if (r < 0) {
+    ret_code = r;
+    goto cleanup;
+  }
+
+  image_snaps = vector<cls::rbd::ImageSnapshotRef>(n, cls::rbd::ImageSnapshotRef());
+
+  ind_snap_name = calc_ind_image_snap_name(group_ioctx.get_id(),
```

Comment:
Can just re-use the group snapshot uuid -- since it's guaranteed to be unique within the group namespace